### PR TITLE
[PBR 2] PBR GL shaders (forward rendering)

### PR DESCRIPTION
**Pull request by bigbike**
## Motivation and Context
This is the 2nd diff in the PBR project.

This diff is to introduce the PBR OpenGL shaders for forward rendering.

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/shaders/pbr.frag [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+  // See here: http://www.thetenthplanet.de/archives/1180
+  vec3 q1 = dFdx(position);
+  vec3 q2 = dFdy(position);
+  vec2 st1 = dFdx(texCoord);
+  vec2 st2 = dFdy(texCoord);
+
+  vec3 N = normalize(normal);
+  vec3 T = normalize(q1 * st2.t - q2 * st1.t);
+  vec3 B = -normalize(cross(N, T));
+  mat3 TBN = mat3(T, B, N);
```
Comment (mosra) [inline src/shaders/pbr.frag:111]: Didn't we agree we'd drop this and render without a normal map instead if there are no tangents (like done with `GenericDrawable` in #837) ? Since, as you said, this caused some weird artifacts in some cases.

**Review by mosra — CHANGES_REQUESTED** [review]
Compared to #837 and #838, which were the right size, I fear there's too little in this PR -- it should have the C++ PbrShader class as well, the GLSL sources alone have no use.

File: src/shaders/pbr.frag [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+// if .w == 0, it means it is a dirctional light, .xyz is the direction;
+// if .w == 1, it means it is a point light, .xyz is the light position;
+// it is NOT put in the Light Structure, simply because we may modify the code
+// so it is computed in the vertex shader.
+uniform vec4 LightDirections[LIGHT_COUNT];
+#endif
+
+// -------------- material, textures ------------------
+struct MaterialData {
+  vec4 baseColor;     // diffuse color, if baseColorTexture exists,
```
Comment (eundersander) [inline src/shaders/pbr.frag:42]: Does the alpha component mean anything?

File: src/shaders/pbr.frag [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+#if defined(NORMAL_TEXTURE) && defined(NORMAL_TEXTURE_SCALE)
+uniform mediump float normalTextureScale
+#ifndef GL_ES
+    = 1.0
+#endif
+    ;
+#endif
+
+// -------------- shader ------------------
+#if defined(NORMAL_TEXTURE)
```
Comment (eundersander) [inline src/shaders/pbr.frag:88]: Does GLSL support #include's ? It seems this normal-mapping logic could be reused in other shaders. It's not specific to PBR.

File: src/shaders/pbr.frag [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+    // radiance
+    vec3 lightRadiance = LightColors[iLight] * attenuation;
+
+    // light source direction: a vector from current position to the light
+    vec3 light = normalize(LightDirections[iLight].xyz -
+                           position * LightDirections[iLight].w);
+
+    finalColor += microfacetModel(baseColor.rgb, metallic, roughness, n, light,
+                                  view, lightRadiance);
+    finalAlpha += baseColor.w;
```
Comment (eundersander) [inline src/shaders/pbr.frag:282]: I don't understand this line.

File: src/shaders/pbr.frag [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+    // attenuation is always 1.0
+    highp float dist =
+        length(LightDirections[iLight].xyz) * LightDirections[iLight].w;
+    // If range is 0 for whatever reason, clamp it to a small value to
+    // avoid a NaN when dist is 0 as well (which is the case for
+    // directional lights).
+    highp float attenuation =
+        clamp(1.0 - pow(dist / max(LightRanges[iLight], 0.0001), 4.0), 0.0,
+              1.0) /
+        (1.0 + dist);
```
Comment (eundersander) [inline src/shaders/pbr.frag:270]: Nit: @mosra  , I saw you did this divide by `(1.0 + dist)` in the phong shader, too. `1 / (1.0 + dist)^2` is not quite the same as `1 / dist^2`. In particular, it behaves very differently for small distances. Isn't this deviating from the standard falloff calculation?

Comment (mosra) [reply src/shaders/pbr.frag:270]: Yeah I spent some time investigating this already ([check out the docs](https://doc.magnum.graphics/magnum/classMagnum_1_1Trade_1_1LightData.html#Trade-LightData-attenuation)) -- If you don't add the 1, it'll explode to infinity at `dist` close to 0, which I don't think is desired :) 

For some reason the glTF spec [doesn't include that in the equation](https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Khronos/KHR_lights_punctual#range-property), but the UE4 calculation (the glTF equation was based on) [has it](http://blog.selfshadow.com/publications/s2013-shading-course/karis/s2013_pbs_epic_notes_v2.pdf):

image

Hmm ... but it should be `(1.0 + dist^2)` not `(1.0 + dist)^2`. I have to check why I did that.

File: src/shaders/pbr.frag (hunk 3 above)
Comment (mosra) [reply src/shaders/pbr.frag:88]: Not really, unless the shader is preprocessed externally first (which is what I'm working on right now, actually). However you can supply multiple source files and do it that way instead.

On the other hand, I don't think it's likely Habitat will have more shaders? As in, the shader will eventually be replaced with one from Magnum instead.

File: src/shaders/pbr.frag (hunk 2 above)
Comment (mosra) [reply src/shaders/pbr.frag:42]: Yep, glTF supports alpha-masked and alpha-blended materials.

Comment (bigbike) [reply src/shaders/pbr.frag:42]: It means a lot.
We need the alpha value if the object is a transparent or semi-transparent object.

File: src/shaders/pbr.frag (hunk 3 above)
Comment (bigbike) [reply src/shaders/pbr.frag:88]: I remembered I thought about this question and googled it a couple of years ago. And I decided not to share the code among shaders.
Take a look:
https://computergraphics.stackexchange.com/questions/100/sharing-code-between-multiple-glsl-shaders

File: src/shaders/pbr.frag (hunk 4 above)
Comment (bigbike) [reply src/shaders/pbr.frag:282]: @eundersander: `.w` should between 0.0 and 1.0. Correct? :)

File: src/shaders/pbr.frag (hunk 3 above)
Comment (mosra) [reply src/shaders/pbr.frag:88]: As magnum handles `#version` itself because of exactly the reasons mentioned in the SO answer, the only remaining valid point from there is

> 2. It's harder to control what needs to be included from within the shader (you need a separate system for this.)

Yes, that's true, but then again you have to prepend all the defines for `LIGHT_COUNT`, `NORMAL_TEXTURE` etc anyway, so having a separate external file doesn't make it *that much* worse.

File: src/shaders/pbr.frag (hunk 5 above)
Comment (bigbike) [reply src/shaders/pbr.frag:270]: @eundersander : You concern sounds reasonable.

About the attenuation:
Here, I followed what was used in the Phong shader in Magnum. When I coded it up, I understand the `1.0` is to avoid division by 0.
I did not dig into it further. But if the paper says it should be `1 + dist^2`, I will change it.

File: src/shaders/pbr.frag (hunk 1 above)
Comment (bigbike) [reply src/shaders/pbr.frag:111]: @mosra : Yes, we had that agreement.

But later I thought it might be good to leave it here as a fallback. It was correct in most cases.
I saw many systems using this trick, for example, here: 
https://github.com/SaschaWillems/Vulkan-glTF-PBR/blob/master/data/shaders/pbr_khr.frag#L125

Yes, true, there are artifacts in some corner cases. If you still feel uncomfortable, I can remove it though.

**Comment by bigbike** [comment]
> Compared to #837 and #838, which were the right size, I fear there's too little in this PR -- it should have the C++ PbrShader class as well, the GLSL sources alone have no use.

I just try to make the reviewer's life easier. I was planning to add the PbrShader classes in the next PR. But since you asked, I will put it in this PR.

File: src/esp/gfx/PbrShader.cpp [hunk 6]
```diff
@@ -0,0 +1,486 @@
...
+
+namespace Mn = Magnum;
+namespace Cr = Corrade;
+
+namespace esp {
+namespace gfx {
+
+namespace {
+enum TextureUnit : uint8_t {
+  BASE_COLOR = 0,
```
Comment (bigbike) [inline src/esp/gfx/PbrShader.cpp:44]: Well, we just discussed. I will change them back to `PascalCase`.

File: src/shaders/pbr.frag (hunk 3 above)
Comment (Skylion007) [reply src/shaders/pbr.frag:88]: @mosra Adopting libshader_rc for Magnum would automatically handle the #INCLUDE preprocessing as well. :)

File: src/shaders/pbr.frag (hunk 5 above)
Comment (bigbike) [reply src/shaders/pbr.frag]: I just updated the code using the `1+dist^2` now. Take a look.

**Comment by bigbike** [comment]
PbrShader class is added in this diff. It is ready to be reviewed.
I hope we can commit it in as soon as possible as other PRs are being blocked by it. Thanks a lot!

File: src/esp/gfx/PbrShader.cpp [hunk 7]
```diff
@@ -0,0 +1,481 @@
...
+// it requires GL4.3 (We stick to GL4.1 for MacOS).
+PbrShader& PbrShader::bindBaseColorTexture(Mn::GL::Texture2D* texture) {
+  if ((flags_ & Flag::BaseColorTexture) && lightCount_ && texture) {
+    texture->bind(TextureUnit::BaseColor);
+  }
+  return *this;
+}
+
+PbrShader& PbrShader::bindRoughnessTexture(Mn::GL::Texture2D* texture) {
+  if ((flags_ & Flag::RoughnessTexture) && lightCount_ && texture) {
```
Comment (mosra) [inline src/esp/gfx/PbrShader.cpp:253]: Are we really not going with references here? What does the coding style guide say about this? Because I remember (mutable) non-nullable references being used quite on a lot of places in Habitat, and using a reference would make it consistent with what Magnum does -- there a pointer is used *only* if it's allowed to be null.

I don't buy the argument that it makes it "easier to use" because MaterialData also use pointers. 

There's an additional concern I have with silent failures -- if the texture is `nullptr` for some reason or the shader doesn't have `RoughnessTexture` enabled for some reason, it will just silently do nothing. In my opinion it neither makes sense to call `bindRoughnessTexture(nullptr)` nor to call `bindRoughnessTexture()` when the shader doesn't have that enabled, so none of that should be allowed:

1. Accept a reference, so if the caller has a pointer for some reason, *the caller* is responsible for dereferencing it. It's not the shader's business to do error handling on behalf of the caller.
2. *Assert* if this function is called and `RoughnessTexture` is not enabled, instead of silently doing a bad render because I don't know, the shader got compiled with packed textures enabled and you're passing separate maps instead. The assert is what Magnum does as well, and from experience I can say it caught quite a lot of mistakes that would take significant time to discover and debug otherwise.

Cc: @aclegg3 @jturner65

File: src/esp/gfx/PbrShader.h [hunk 8]
```diff
@@ -0,0 +1,453 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+#ifndef ESP_GFX_PBRSHADER_H_
+#define ESP_GFX_PBRSHADER_H_
+
+#include <initializer_list>
+#include <vector>
```
Comment (mosra) [inline src/esp/gfx/PbrShader.h:9]: Unneeded include.

File: src/esp/gfx/PbrShader.h [hunk 9]
```diff
@@ -0,0 +1,453 @@
...
+
+#include <Corrade/Containers/ArrayView.h>
+#include <Corrade/Containers/EnumSet.h>
+#include <Magnum/GL/AbstractShaderProgram.h>
+#include <Magnum/Math/Matrix4.h>
```
Comment (mosra) [inline src/esp/gfx/PbrShader.h:14]: *Technically* you don't need this include either as there's a forward declaration that's enough. 

OTOH, if *do* you want to have the full definition here, you'll need `Matrix3.h` as well for `setTextureMatrix()`.

**Review by mosra — COMMENTED** [review]
Eh, sorry, I thought I'd just approve but more things cropped up.

Will reply to the remaining threads separately.

File: src/shaders/pbr.frag (hunk 4 above)
Comment (mosra) [reply src/shaders/pbr.frag:281]: Eric means it's an useless operation, you could as well do just 

```glsl
fragmentColor += vec4(finalColor, baseColor.a);
```

below and don't calculate any `finalAlpha`.

File: src/shaders/pbr.frag (hunk 3 above)
Comment (mosra) [reply src/shaders/pbr.frag:88]: @Skylion007 Yeah no :P Includes are supported directly by glslang itself, it's just not exposed in the `glslangValidator` command-line tool for whatever reason.

Which means, at this point, I don't see any benefits of using shaderc over glslang + spirv-tools directly, it would only add another hard-to-build dependency to the pipeline and further expand the surface for potential bugs.

File: src/shaders/pbr.frag (hunk 1 above)
Comment (mosra) [reply src/shaders/pbr.frag:111]: It's easy to add tangents in an offline processing step (via Blender, for example) and most models already have them. The only PBR models that *didn't* have those were the ones exported from Blender without the Tangents option checked. 

This additional code creates another shader variant we need to test and maintain, so I'd vote for removing it -- the best code is code that's not written :P

File: src/shaders/pbr.frag (hunk 5 above)
Comment (mosra) [reply src/shaders/pbr.frag]: Okay, then I'll do the same change in Phong. Will get back to you if I discover any issues with that.

File: src/esp/gfx/PbrShader.cpp [hunk 10]
```diff
@@ -0,0 +1,481 @@
...
+    attributeLocationsStream << Cr::Utility::formatString(
+        "#define ATTRIBUTE_LOCATION_TANGENT4 {}\n", Tangent4::Location);
+  }
+  auto isTextured = [&]() {
+    return (flags_ &
+            (Flag::BaseColorTexture | Flag::RoughnessTexture |
+             Flag::MetallicTexture | Flag::NormalTexture |
+             Flag::EmissiveTexture | Flag::NoneRoughnessMetallicTexture |
+             Flag::OcclusionRoughnessMetallicTexture));
+  };
```
[the remark this answers is not in the document]
Comment (bigbike) [reply src/esp/gfx/PbrShader.cpp:110]: Yes, it makes sense.

File: src/esp/gfx/PbrShader.cpp (hunk 7 above)
Comment (bigbike) [reply src/esp/gfx/PbrShader.cpp:253]: > I don't buy the argument that it makes it "easier to use" because MaterialData also use pointers.

Yes, for the reason "you do not buy". It is much more convenient to use pointer here. In the render() in PbrDrawable (which is not in this PR), I can use the method chaining to call them, and do not need to care about if the texture exists, or if the shader is initialized with this feature.
```
      .bindBaseColorTexture(materialData_->baseColorTexture)
      .bindRoughnessTexture(materialData_->roughnessTexture)
      .bindMetallicTexture(materialData_->metallicTexture)
      .bindNoneRoughnessMetallicTexture(
          materialData_->noneRoughnessMetallicTexture)
      .bindOcclusionRoughnessMetallicTexture(
          materialData_->occlusionRoughnessMetallicTexture)
      .bindNormalTexture(materialData_->normalTexture)
      .bindEmissiveTexture(materialData_->emissiveTexture)
```
If you change it to the reference (which I tried indeed) you will have a lot of if clause here.

This issue actually reflects different view between us in terms of the robustness of the code.
- I tried to keep the system running, not being interrupted (terminated) due to developers' careless mistakes (NON-fatal errors, e.g., there is actually no roughness texture, but she bound it), and I hided the sanity checks within my APIs to make sure the system will not crash due to these "mistakes". I choose a more tolerant way, and provide easy-to-use APIs (e.g., here you can use method chaining, without any if clauses, and you do not have to remember if you activate corresponding feature or not). "Lazy" developers are going to like it. :)

- You used a more active way to encourage the developer "fail" earlier. As you mentioned, " I can say it caught quite a lot of mistakes that would take significant time to discover and debug otherwise." which I agree. 

In short, the difference is that you leave such overhead (sanity check, dereferencing, assert etc.) to the developer, but I do not. No right or wrong here. Just different style.

I just explain to let you know my view. And I do not think one is more appealing than the other one.

I can change to the way you suggested. I am OK with it.

File: src/esp/gfx/PbrShader.cpp [hunk 11]
```diff
@@ -0,0 +1,481 @@
...
+  }
+  return *this;
+}
+
+PbrShader& PbrShader::setMVPMatrix(const Mn::Matrix4& matrix) {
+  setUniform(mvpMatrixUniform_, matrix);
+  return *this;
+}
+
+PbrShader& PbrShader::setNormalMatrix(const Mn::Matrix3& matrix) {
```
[the remark this answers is not in the document]
Comment (bigbike) [reply src/esp/gfx/PbrShader.cpp:302]: Oh, good to know!
Not sure if you particularly mentioned it in your wiki. If you did not, I recommend you to do that.
I was thinking `Matrix3` is just a convenient `typedef` on top of the `Matrix3x3`.

**Review by bigbike — COMMENTED** [review]
Thanks a lot @mosra. I saw many constructive feedback for you. I really appreciated it.

File: src/shaders/pbr.vert [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+#endif
+
+// ------------ uniform ----------------------
+uniform highp mat4 ModelViewMatrix;
+uniform highp mat3 NormalMatrix;  // inverse transpose of 3x3 modelview matrix
+uniform mat4 MVP;
+
+#ifdef TEXTURE_TRANSFORMATION
+uniform highp mat3 textureMatrix
+#ifndef GL_ES
```
[the remark this answers is not in the document]
Comment (mosra) [reply src/shaders/pbr.vert:34]: Okay, then this `textureMatrix` should be `TextureMatrix` instead :)

File: src/shaders/pbr.frag (hunk 4 above)
Comment (bigbike) [reply src/shaders/pbr.frag:281]: Oh, OK, I see.
The add should happen outside the for loop since baseColor.a never changes.
fragmentColor += vec4(finalColor, baseColor.a);

File: src/shaders/pbr.vert (hunk 12 above)
[the remark this answers is not in the document]
Comment (bigbike) [reply src/shaders/pbr.vert:34]: will do.

File: src/shaders/pbr.vert [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+#if defined(NORMAL_TEXTURE) && defined(PRECOMPUTED_TANGENT)
+  tangent = normalize(NormalMatrix * vec3(vertexTangent));
+  // Gram–Schmidt
+  tangent = normalize(tangent - dot(tangent, normal) * normal);
+  biTangent = normalize(cross(normal, tangent) * vertexTangent.w);
+  // later in .frag, TBN will transform the normal perturbation
+  // (read from normal map) from tangent space to camera space
+#endif
+
+  gl_Position = MVP * vertexPosition;
```
[the remark this answers is not in the document]
Comment (mosra) [reply src/shaders/pbr.vert:61]: To clarify, I have nothing against the notion of a model-view-projection matrix. It's just that I find it more logical to supply a separate `modelview` and `projection` parts than `modelview` and `modelviewprojection`, where the same information is encoded twice.

And as a bonus it'll also remove one unnecessary matrix multiplication in the Drawable implementation, if I'm not mistaken.

File: src/esp/gfx/PbrShader.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (mosra) [reply src/esp/gfx/PbrShader.cpp:302]: Good point. [It is mentioned at one place](https://doc.magnum.graphics/magnum/namespaceMagnum.html#a7ee135be46a327f352d01287579e2acf), but I'll put that information elsewhere as well.

File: src/esp/gfx/PbrShader.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (bigbike) [reply src/esp/gfx/PbrShader.cpp:110]: by the way, how to convert:
```
flags_ &
            (Flag::BaseColorTexture | Flag::RoughnessTexture |
             Flag::MetallicTexture | Flag::NormalTexture |
             Flag::EmissiveTexture | Flag::NoneRoughnessMetallicTexture |
             Flag::OcclusionRoughnessMetallicTexture);
```
to a `bool`. (currently I am using auto).

[the remark this answers is not in the document]
Comment (mosra) [reply src/esp/gfx/PbrShader.cpp:110]: It has an [explicit operator bool](https://doc.magnum.graphics/corrade/classCorrade_1_1Containers_1_1EnumSet.html#aaa72f9d5d7ab6e51d66d3698206e7cc9), so either `!!(thiswholething)` or `bool(thiswholething)`.

Implicit bool conversion would make it convertible to an int, which is not desired from a type safety perspective.

File: src/esp/gfx/PbrShader.cpp (hunk 7 above)
Comment (mosra) [reply src/esp/gfx/PbrShader.cpp:253]: Funnily enough right now I'm fighting `glslang` which is built with this "fail silently" approach and it costs me an immense amount of time because a lot of the time it just does nothing and i don't know *why* :laughing: 

Yep, It's a matter of perspective. Could we get opinions from others here? I don't want to push my view if I would the only one who wants it that way.

File: src/esp/gfx/PbrShader.h (hunk 9 above)
Comment (bigbike) [reply src/esp/gfx/PbrShader.h:14]: I do not need either of them in the .h file.

File: src/esp/gfx/PbrShader.cpp (hunk 7 above)
Comment (bigbike) [reply src/esp/gfx/PbrShader.cpp]: I changed them to use the reference here, and also added assertions within the API to make it fail immediately in illegal usage cases. It is worth if "it caught quite a lot of mistakes that would take significant time to discover and debug otherwise." We can continue the conversation on this topic in the slack channel.

This PR is out for quite a long time, and I hope there is no more further delays.

**Review by mosra — APPROVED** [review]
:tada:

**Review by Skylion007 — APPROVED** [review]

**Comment by Skylion007** [comment]
@bigbike LGTM lets get this merged! :)

File: src/esp/gfx/PbrShader.cpp (hunk 7 above)
Comment (bigbike) [reply src/esp/gfx/PbrShader.cpp]: @eundersander : Here is the discussion on the design philosophy.
